src/librustc_middle/ty/mod.rs

```diff
@@ -1160,14 +1156,14 @@ impl<'tcx> GenericPredicates<'tcx> {
 }
 
 #[derive(Clone, Copy, PartialEq, Eq, Hash, RustcEncodable, RustcDecodable, Lift)]
-#[derive(HashStable, TypeFoldable)]
+#[derive(HashStable)]
 pub struct Predicate<'tcx> {
-    kind: PredicateKind<'tcx>,
+    kind: &'tcx PredicateKind<'tcx>,
 }
 
 impl Predicate<'tcx> {
     pub fn kind(&self) -> PredicateKind<'tcx> {
```

Comment:
so we had talked about adding a follow-up commit to make

`kind(&self) -> &PrediceKind<'tcx>`

Reply:
We did :sparkles:

IMO we should replace `Predicate` with `type Predicate<'tcx> = &'tcx PredicateKind<'tcx>`, even if we have to manually implement `TypeFoldable` and friends in this case.